pkg/common/scheduler/statefulset/autoscaler.go

```diff
@@ -155,7 +162,60 @@ func (a *autoscaler) doautoscale(ctx context.Context, attemptScaleDown bool, pen
 			a.logger.Errorw("updating scale subresource failed", zap.Error(err))
 			return err
 		}
+	} else {
+		// since the number of replicas hasn't change, take the opportunity to
+		// compact the vreplicas
+		a.mayCompact(state)
 	}
 
 	return nil
 }
+
+func (a *autoscaler) mayCompact(s *state) {
+	// when there is only one pod there is nothing to move!
+	if s.lastOrdinal < 1 {
+		return
+	}
+
+	if s.schedulerPolicy == MAXFILLUP {
+		// Determine if there is enough free capacity to
+		// move all vreplicas placed in the last pod to pods with a lower ordinal
+		freeCapacity := s.freeCapacity() - s.Free(s.lastOrdinal)
+		usedInLastPod := s.capacity - s.Free(s.lastOrdinal)
+
+		if freeCapacity >= usedInLastPod {
+			err := a.compact(s)
+			if err != nil {
+				a.logger.Errorw("vreplicas compaction failed", zap.Error(err))
+			}
+		}
+
+		// only do 1 replica at a time to avoid overloading the scheduler with too many
+		// rescheduling requests.
+	}
+}
+
+func (a *autoscaler) compact(s *state) error {
+	a.logger.Info("compacting vreplicas")
+	vpods, err := a.vpodLister()
+	if err != nil {
+		return err
+	}
+
+	for _, vpod := range vpods {
+		placements := vpod.GetPlacements()
+		for _, placement := range placements {
+			ordinal := ordinalFromPodName(placement.PodName)
+
+			if ordinal == s.lastOrdinal {
+				a.logger.Infow("evicting vreplica(s)", zap.String("podname", placement.PodName), zap.Int("vreplicas", int(placement.VReplicas)))
```

Comment:
please also log the name of the Kafka source being evicted as well.

Reply:
done!